src/ipc/local/LocalIPCMessage.cpp

```diff
@@ -96,6 +86,17 @@ void LocalIPCMessage::copyRequestMessage(const LocalIPCMessage &other)
     }
 }
 
+QList<QVariant> LocalIPCMessage::arguments() const
+{
+    return m_data.m_arguments;
+}
+
+LocalIPCMessage &LocalIPCMessage::operator<<(const QVariant &arg)
```

Comment:
no conversion to QVariant is expected for local IPC

Reply:
It is the Qt standard way of treating different types generically. As opposed to strange pattern of facelift::structure and collecting all fields in a std::tuples and its overhead one should have used QVariant when working with Qt.
The next step would be to remove the std::tuple.

Reply:
QVariant provides a way to handle different types generically at runtime. We do not need that here, and we do not want to pay the price of the overhead which the usage of QVariant introduces. All types are known at build time, and we want to benefit from C++'s type safety to minimize the potential runtime failures.

Reply:
std::tuple overhead:
The bottle-neck is the construction of std::tuples at facelift::structure, as I already wrote the benchmark shows replacing std::tuple with some more efficient types like [tao::tuple](https://github.com/taocpp/tuple/pulls) proves ~15% speed gain. So how much we would gain by not having any tuple at all?!
Feel free to do more benchmarks.
Rationale:
Obviously QVariant creates some overhead at this phase but it is absolutely necessary to treat the IPCMessages generically, given that QtDBus is laid out on QVariants.
Type safety:
Current implementation of serializing values to input/Output payload (QByteArray) when sending local/DBus messages doesn't provide any type safety that QVariant would take away. In contrast using QtDBus we have clearer methods/struct signature (as oppose to ByteArray) for cross-referencing.

Reply:
> std::tuple overhead:
> The bottle-neck is the construction of std::tuples at facelift::structure, as I already wrote the benchmark shows replacing std::tuple with some more efficient types like [tao::tuple](https://github.com/taocpp/tuple/pulls) proves ~15% speed gain. So how much we would gain by not having any tuple at all?!

Where can I find that benchmark ? AFAIS, the main goal of tao::tuple is to optimize compilation time.

> Feel free to do more benchmarks.
> Rationale:
> Obviously QVariant creates some overhead at this phase but it is absolutely necessary to treat the IPCMessages generically, given that QtDBus is laid out on QVariants.

Use QVariant where it is needed, which means in the DBus IPC classes, not anywhere else.

> Type safety:
> Current implementation of serializing values to input/Output payload (QByteArray) when sending local/DBus messages doesn't provide any type safety that QVariant would take away. In contrast using QtDBus we have clearer methods/struct signature (as oppose to ByteArray) for cross-referencing.

Reply:
You need to create facelift specific benchmarks, I created ones with function calls with different range of structs and size.
Please see my other comment about QVariant and facelift payload.